x-pack/plugins/infra/server/routes/services/index.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import Boom from '@hapi/boom';
+import {
+  GetServicesRequestQueryRT,
+  GetServicesRequestQuery,
+  ServicesAPIResponseRT,
+} from '../../../common/http_api/host_details';
+import { InfraBackendLibs } from '../../lib/infra_types';
+import { getServices } from '../../lib/host_details/get_services';
+import { validateStringAssetFilters } from './lib/utils';
+import { createSearchClient } from '../../lib/create_search_client';
+import { buildRouteValidationWithExcess } from '../../utils/route_validation';
+
+export const initServicesRoute = (libs: InfraBackendLibs) => {
+  const validate = buildRouteValidationWithExcess(GetServicesRequestQueryRT);
+
+  const { framework } = libs;
+  framework.registerRoute<unknown, GetServicesRequestQuery, unknown>(
+    {
+      method: 'get',
+      path: '/api/infra/services',
+      validate: {
+        query: (q, res) => {
+          const [invalidResponse, parsedFilters] = validateStringAssetFilters(q, res);
+          if (invalidResponse) {
+            return invalidResponse;
+          }
+          if (parsedFilters) {
+            q.validatedFilters = parsedFilters;
+          }
+          return validate(q, res);
+        },
+      },
+    },
+    async (requestContext, request, response) => {
+      const [{ savedObjects }] = await libs.getStartServices();
+      const { from, to, size = 10, validatedFilters } = request.query;
+
+      try {
+        if (!validatedFilters) {
+          throw Boom.badRequest('Invalid filters');
+        }
+        const client = createSearchClient(requestContext, framework, request);
+        const soClient = savedObjects.getScopedClient(request);
+        const apmIndices = await libs.getApmIndices(soClient);
+        const services = await getServices(client, apmIndices, {
+          from,
+          to,
+          size,
+          filters: validatedFilters,
+        });
+        return response.ok({
+          body: ServicesAPIResponseRT.encode(services),
+        });
+      } catch (err) {
+        if (Boom.isBoom(err)) {
+          return response.customError({
+            statusCode: err.output.statusCode,
+            body: { message: err.output.payload.message },
+          });
+        }
+
+        return response.customError({
+          statusCode: err.statusCode ?? 500,
+          body: {
+            message: err.message ?? 'An unexpected error occurred',
+          },
+        });
+      }
+    }
```

Comment:
There is a lot of boilerplate here and it's hard to see what it has to do with this route. Mostly the validation and error handling looks very generic. Shouldn't this be handled by the framework?

Reply:
I cleaned this up.  When trying to validate a strict type of allowed filters it made things a bit more complicated.  TS can't infer that validatedFilters exist which is a different type than the filters param which is a string.  Since it definitely does exist or it would fail in validateStringAssetFilters, I've used a type assertion.